### DISCUSSION TITLE
Possible to create a tb top and use configDB before uvm_test_top is created?

### DISCUSSION THREAD

**Discussion by c-bjerrum** [pyuvm/pyuvm#139] [Q&A]
Is it possible to create a tb top class or something equivalent?  

I would like to instantiate my interface in the tb top and then pass it to the configDB, as I would like to create the interface outside the test.  

I was thinking that I might be able to use the `uvm_root` class to create a tb top in which I can set() in the configDB. But this gives an error: `pyuvm.error_classes.UVMConfigItemNotFound: "uvm_test_top" is not in ConfigDB().`

    @cocotb.test()
    async def start(_):
        await tb_top().run_test("test")

    class tb_top(uvm_root):
         """Top module for the tb"""
        def build_phase(self):
            ConfigDB().set(None, "*","MSG","This is a message")
        
    class test(uvm_test):
        def build_phase(self):
            msg = ConfigDB().get(self,"","MSG")
            print(msg)

Any ideas how to solve this?

**Comment by raysalemi** [comment]
Have you considered using the `@pyuvm.test()` decorator on your tests?

It's not clear to me that you've done anything to tell Python that `tb_top` is in the UVM hierarchy.

**Reply by c-bjerrum** [reply]
Thank you for your answer. 

It was decision not adding the `@pyuvm.test()` decorator and instead using the `uvm_root().run_test(<testname>)` as I would like to control the test to run from a makefile. I found an example on git yesterday, which I think you have created, where the same thing was done, but I have not been able to find it again. 

I agree that the `tb_top` is not added to the hierarchy which causes problems. I assume that I could work around the problem like this instead. 

    @pyuvm.test()
    class tb_top(uvm_component):
        def build_phase(self):
            self.sdt_if = cl_sdt_interface("sdt_if")
            ConfigDB().set(self, "*","sdt_if","sdt_if")
        
            test("test",self)
